freeipa/src/main/java/com/sequenceiq/freeipa/service/freeipa/flow/FreeIpaInstallService.java

```diff
@@ -44,6 +56,13 @@ public void installFreeIpa(Long stackId) throws CloudbreakOrchestratorException
 
     private void installFreeIpa(Long stackId, Stack stack, List<GatewayConfig> gatewayConfigs, Set<Node> allNodes) throws CloudbreakOrchestratorException {
         GatewayConfig primaryGatewayConfig = gatewayConfigService.getPrimaryGatewayConfig(stack);
+        List<RecipeModel> recipes = freeIpaRecipeService.getRecipes(stack.getRecipes());
+        LOGGER.info("Recipes for stack: {}", recipes);
+        Map<String, List<RecipeModel>> recipeMap = stack.getInstanceGroups().stream().map(InstanceGroup::getGroupName)
+                .collect(Collectors.toMap(instanceGroup -> instanceGroup, instanceGroup -> recipes));
+        hostOrchestrator.uploadRecipes(gatewayConfigs, recipeMap, new StackBasedExitCriteriaModel(stackId));
+        hostOrchestrator.preClusterManagerStartRecipes(primaryGatewayConfig, allNodes, new StackBasedExitCriteriaModel(stackId));
         hostOrchestrator.installFreeIpa(primaryGatewayConfig, gatewayConfigs, allNodes, new StackBasedExitCriteriaModel(stackId));
+        hostOrchestrator.postClusterManagerStartRecipes(primaryGatewayConfig, allNodes, new StackBasedExitCriteriaModel(stackId));
```

Comment:
I'm not sure why is 2 hook is necessary, one here and one in `FreeIpaPostInstallService`

Reply:
Post-start recipes and post-install recipes are different and they run at a different phases of provisioning. Or maybe I don't really understand this.

Reply:
so you have placed 2 post recipes into the process:
- one right after the salt calls, which mostly pure freeipa installation (with replicas) and a minimal extra configuration (which needs to be done via salt)
- and one after the post install, which configures the replication topology, the admin user, password policy etc

I'm not sure if the first one is necessary, and we can't get away with only one post recipe. Is there any use case or business justification for it?

Reply:
I don't know if there is any business justification, but we can remove it later if it does not make sense. For now I would keep both of them because I don't want to limit which recipes can be attached to freeipa, it would increase the complexity. I must rename recipe types in my next PR, then we will think about if it has any business value to having both types.

Reply:
they could run at the same time in case of freeipa :)

Reply:
I removed hostOrchestrator.postClusterManagerStartRecipes call! Thanks!